### PR TITLE
[Tiered Storage] Fix read tiered storage ledger data again

### PR DESCRIPTION
**Pull request by gaoran10**
### Motivation

Currently, the `BlobStoreBackedReadHandleImpl` couldn't read the ledger data again, if read to the end of the data stream, it can't reset the reader index. So the `ReadHandle`(BlobStoreBackedReadHandleImpl) of the ledger couldn't be reused.

**How to reproduce**
1. Create a consumer and consume offloaded data(don't ack messages).
2. Restart the consumer and read data from the earliest position.
3. We'll meet this issue.

**error logs**
```
2021-09-19T00:56:00,806+0800 [TestNG-method=testOffloadAndRead-1:org.apache.bookkeeper.mledger.offload.jcloud.impl.BlobStoreManagedLedgerOffloaderTest@146] ERROR org.apache.bookkeeper.mledger.offload.jcloud.impl.BlobStoreManagedLedgerOffloaderTest - Failed to read offload data again.
org.apache.bookkeeper.client.api.BKException: Unexpected condition
	at org.apache.bookkeeper.client.api.BKException.lambda$static$0(BKException.java:40) ~[bookkeeper-server-4.14.2.jar:4.14.2]
	at org.apache.bookkeeper.common.concurrent.FutureUtils.result(FutureUtils.java:77) ~[bookkeeper-common-4.14.2.jar:4.14.2]
	at org.apache.bookkeeper.client.api.ReadHandle.read(ReadHandle.java:58) ~[bookkeeper-server-4.14.2.jar:4.14.2]
	at org.apache.bookkeeper.mledger.offload.jcloud.impl.BlobStoreManagedLedgerOffloaderTest.testOffloadAndRead(BlobStoreManagedLedgerOffloaderTest.java:144) [test-classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_261]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_261]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_261]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_261]
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132) [testng-7.3.0.jar:?]
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45) [testng-7.3.0.jar:?]
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73) [testng-7.3.0.jar:?]
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11) [testng-7.3.0.jar:?]
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266) [?:1.8.0_261]
	at java.util.concurrent.FutureTask.run(FutureTask.java) [?:1.8.0_261]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_261]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_261]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_261]
Caused by: java.io.EOFException
	at java.io.DataInputStream.readInt(DataInputStream.java:392) ~[?:1.8.0_261]
	at org.apache.bookkeeper.mledger.offload.jcloud.impl.BlobStoreBackedReadHandleImpl.lambda$readAsync$1(BlobStoreBackedReadHandleImpl.java:111) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_261]
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125) ~[guava-30.1-jre.jar:?]
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69) ~[guava-30.1-jre.jar:?]
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78) ~[guava-30.1-jre.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_261]
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266) ~[?:1.8.0_261]
	at java.util.concurrent.FutureTask.run(FutureTask.java) ~[?:1.8.0_261]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_261]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_261]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_261]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_261]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.67.Final.jar:4.1.67.Final]
	... 1 more

```

### Modifications

Reset the buffer reader index if the reading progress reaches the end of the data stream.

### Verifying this change

Modify the test to verify reading tiered storage data again.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

- [ ] doc-required 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] no-need-doc 
  
  (Please explain why)
  
- [ ] doc 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Review by hangc0276 — APPROVED** [review]
Nice catch!

**Comment by eolivelli** [comment]
@michaeljmarshall would you mind to take a look please?

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java
Comment (eolivelli) [inline tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java:95]: Is this block to be moved below inside the lamdba ?
It looks to me that we are not handling correctly concurrency

**Review by eolivelli — CHANGES_REQUESTED** [review]
While I believe that this could fix the problem I am not sure this is a good way.
Why we have to add those two methods?
Isn't it possible to hide that implementation details inside the InputStream implementation?

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java
Comment (michaeljmarshall) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java:95]: Yes, I agree this is not thread safe. The calling thread for `readAsync` could be any thread. The `executor` is from an Ordered Executor using the ledger id as the mod. Since all updates to the buffer are run by the same `executor`, the current paradigm is thread safe. We should only update buffer state from within a lambda on this class's executor service.

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java
Comment (michaeljmarshall) [inline tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java:143]: Nit: what do you think about wrapping this debug statement in a conditional to prevent unnecessary boxing as well as unnecessary computation?

**Review by michaeljmarshall — CHANGES_REQUESTED** [review]
@gaoran10 - can you help me understand the motivation for this change a bit more? When does the `readAsync` method get called multiple times on the same buffer? Thanks.

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java
Comment (gaoran10) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java:95]: Maybe the method `readAsync` of the class `BlobStoreBackedReadHandleImpl` is not thread-safe, because the seek method will also reset the reader index of the `ByteBuf`.

**Comment by gaoran10** [comment]
@michaeljmarshall If a consumer consumes the topic and didn't ack messages. When the consumer re-consume the topic from the earliest position, the `ReadHandle` need to read data from the head of the data stream. But if the reader index at the end of the `ByteBuf` it couldn't read an int to check the `EntryId` and it will meet this issue.

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java
Comment (gaoran10) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java:143]: There is no extra method in `log.debug(xxx)`, the method `log.debug` will check the log level first and it seems that there is no extra cost.

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java
Comment (michaeljmarshall) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java:95]: My point is not that `readAsync` is not thread-safe, but rather that updating the buffer's internal state in the way this PR proposes is not thread-safe. I think we should only update the buffer's state from within the lambda to ensure that the buffer is only ever updated by a single thread. Note that the seek method updates the buffer from the executor's single thread in the lambda.

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java
Comment (michaeljmarshall) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java:143]: `this.buffer.readableBytes()` is the method I was referring to. It will be computed before `log.debug` is called. Also, before `log.debug` is called, the `cursor` and `objectLen` variables, which are both `long`, will be converted into `Long` causing a heap allocation. Since this `readAsync` method is on a critical path for reading data from tiered storage, I think these minor optimizations are valuable.

This was also discussed here: https://github.com/apache/pulsar/pull/11880#discussion_r700384425.

**Comment by michaeljmarshall** [comment]
> @michaeljmarshall If a consumer consumes a topic and didn't ack messages. When the consumer consumes the topic from the earliest position again, the `ReadHandle` needs to read data from the head of the data stream. But if the reader index at the end of the `ByteBuf` it couldn't read an int to check the `EntryId` and it will meet this issue.
> 
> I add a section to describe how to reproduce this problem. PTAL

Thank you for the additional explanation, @gaoran10.

**Review by jiazhai — APPROVED** [review]

**Comment by jiazhai** [comment]
+1. nice catch and fix 👍

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java
Comment (eolivelli) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java:143]: in Pulsar we have this convention to always add log guards.
please add if (log.isDebugEnable()).

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java
Comment (eolivelli) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java:95]: exactly, we must move the code inside the lambda

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java
Comment (gaoran10) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedInputStreamImpl.java:143]: Ok, I'll fix this.

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java
Comment (gaoran10) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java]: Sorry, I misunderstand the problem, nice catch, I'll check the inputStream remaining data in lambda.

**Comment by gaoran10** [comment]
I check the inputStream remaining data in lambda. PLTA @michaeljmarshall @eolivelli

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by eolivelli** [comment]
@michaeljmarshall can you please give your final word ?

**Comment by michaeljmarshall** [comment]
@gaoran10 @eolivelli @zymap @hangc0276 @jiazhai - I think there are conflicts between this PR and https://github.com/apache/pulsar/pull/12123. Specifically, it looks like they both try to prevent `dataStream.readInt()` from reaching the end of the stream. This PR proposes always setting the the `readerIndex` to the start of the stream if there is data. https://github.com/apache/pulsar/pull/12123 uses seek to achieve a similar end here: https://github.com/apache/pulsar/pull/12123/files#diff-03c8f778b41079e0dc3163a9374e4d291bc6544d84e74b6ab419b0e383032e44R118-R125.

I'm not sure what the "right" solution is here, but I think we should try to solve this problem before merging either PR.

**Comment by gaoran10** [comment]
I think the solution in PR https://github.com/apache/pulsar/pull/12095 is better, it will seek the reader index of the inputstream to the suitable position first. I close this PR first.

**Pull request closed without merging**
